### PR TITLE
Added Kconfig choice for c library impl. + corrected newlib dependencies 

### LINKED ISSUE #18305: Native Posix target can not use features with newlib dependencies
**Describe the bug**
In case you are building for native posix target the newlib C library can not be selected but some features have a hard dependency towards newlib even when provided by OS libc implementation.

**To Reproduce**
One example more issues can be found in code base:
1. west build -b native_posix samples/hello_world -- -DCONFIG_CPLUSPLUS=y -DCONFIG_EXCEPTIONS=y

```
warning: EXCEPTIONS (defined at subsys/cpp/Kconfig:55) was assigned the value 'y' but got the value
'n'. You can check symbol information (including dependencies) in the 'menuconfig' interface (see
the Application Development Primer section of the manual), or in the Kconfig reference at
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_EXCEPTIONS.html (which is updated
regularly from the master branch). See the 'Setting configuration values' section of the Board
Porting Guide as well.
```

**Expected behavior**
You should be able to select all features that only depends on newlib for the native posix target.

**Impact**
Some features can not be enabled on native posix target

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: gcc 7.4
 - Commit SHA: 4ea3c9560ac3823ea4a5f1a79e4661a499941b02

**Proposed Solution**

Add hidden Kconfig symbol `HAS_FULL_LIBC`, which is set both for native posix and newlib, and use this symbol in `depends on`.

**ALternative Solution**

Add `ARCH_POSIX` symbol to all `depends on` where newlib is used

### PR DESCRIPTION
**Pull request by vanwinkeljan**
Added a Kconfig choice to select c library implementation and updated the dependencies towards newlib through the code base.

Fixes #18305

### CODE REVIEW

File: samples/net/updatehub/prj.conf [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Turn on the updatehub
 CONFIG_UPDATEHUB=y
+CONFIG_NEWLIB_LIBC=y
```
Comment (pabigot) [inline samples/net/updatehub/prj.conf]: Why is this change needed?

File: lib/libc/Kconfig [hunk 2]
```diff
@@ -6,19 +6,41 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+
```
Comment (pabigot) [inline lib/libc/Kconfig]: I don't know if we have style guidelines for this, but from the diffs it doesn't appear that double line separation is used consistently here.  I'd personally rather see single blank separation consistently, as it is in the current master version of this file.

File: lib/libc/Kconfig [hunk 3]
```diff
@@ -6,19 +6,41 @@
...
-depends on !NATIVE_APPLICATION
+
+choice LIB_C_IMPLEMENTATION
+	prompt "C Library Implementation"
+	default MINIMAL_LIBC
+
+config MINIMAL_LIBC
+	bool "Minimal c library"
+	depends on !NATIVE_APPLICATION
+	depends on !SOC_SERIES_STM32H7X
```
Comment (pabigot) [inline lib/libc/Kconfig]: I would prefer to see the STM32H7X changes pulled out to a separate commit.

File: lib/libc/Kconfig [hunk 4]
```diff
@@ -6,19 +6,41 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+
 menu "C Library"
-depends on !NATIVE_APPLICATION
+
+choice LIB_C_IMPLEMENTATION
```
Comment (pabigot) [inline lib/libc/Kconfig]: `LIBC_IMPLEMENTATION` for consistency with the other symbols?

**Review by pabigot — CHANGES_REQUESTED** [review]
This is an excellent proposal; I've been unhappy that all the C++ code introduced a dependency on `NEWLIB_LIBC` when it really just needs a "real" libc.

I would like to see a small refactoring of the commits:
* One that introduces and supports `LIBC_IMPLEMENTATION` (not `LIB_C_IMPLEMENTATION`)
* One that switches `NEWLIB_LIBC` to `!MINIMAL_LIBC` (and reverse) everywhere and does nothing else (rather than split into two commits);
* One that introduces any new specific settings for STM32 and SimpleLink (rather than having them hidden in global changes not specific to either platform).

File: samples/net/updatehub/prj.conf (hunk 1 above)
Comment (vanwinkeljan) [reply samples/net/updatehub/prj.conf]: This is need as the `select NEWLIB_LIBC` in `lib/updatehub/Kconfig` has been changed to `depends on !MINIMAL_LIBC`.

The reason behind the Kconfig change is that `select on` is not allowed on config symbols in a choice.

**Comment by vanwinkeljan** [comment]
@pabigot 

> * One that switches NEWLIB_LIBC to !MINIMAL_LIBC (and reverse) everywhere and does nothing else (rather than split into two commits);

Main reason for the two commits is that one just change the `depends on NEWLIB_LIBC` to `depends on !MINIMAL_LIBC` (and variants), the other changes `select NEWLIB_LIB` into `depends on !MINIMAL_LIBC` and asuche changes behaviour

File: lib/libc/Kconfig (hunk 3 above)
Comment (vanwinkeljan) [reply lib/libc/Kconfig]: done

File: drivers/counter/Kconfig.stm32_rtc [hunk 5]
```diff
@@ -8,11 +8,11 @@
 menuconfig COUNTER_RTC_STM32
 	bool "STM32 Counter RTC driver"
 	depends on SOC_FAMILY_STM32
+	depends on !MINIMAL_LIBC
 	select USE_STM32_LL_RTC
 	select USE_STM32_LL_PWR
 	select USE_STM32_LL_RCC
 	select USE_STM32_LL_EXTI
-	select NEWLIB_LIBC
```
Comment (aescolar) [inline drivers/counter/Kconfig.stm32_rtc]: I hope removing these `select`s will not cause too many issues to users..
Including these `select` changes would discourage from merging during freeze even though this PR includes a bugfix for another issue.
Would it make sense to separate the `select` changes commit into a separate PR for after 2.0?
(So just leaving this PR fixing only the bug, and not disturbing other options behavior, including the STM thingy, until after the release)
(The change of `depends on NEWLIB_LIBC` => `depends on !MINIMAL_LIBC` should be safe for applying now).

File: lib/libc/Kconfig [hunk 6]
```diff
@@ -7,18 +7,39 @@
...
 
 config NEWLIB_LIBC
-	bool "Build with newlib c library"
+	bool "Newlib c library"
+	depends on !NATIVE_APPLICATION
 	help
 	  Build with newlib library. The newlib library is expected to be
 	  part of the SDK in this case.
 
+config NATIVE_LIBC
```
Comment (aescolar) [inline lib/libc/Kconfig]: Instead of this can we have a
`config EXTERNAL_LIBC` (or USER_PROVIDED_LIBC or whatever other name) to indicate neither of the 2 libc's provided by Zephyr is used?
That way we cover all 3rd party libCs. (There doesn't seem to be a need, at least yet, to differentiate the host libC from any other 3rd party libC)

Note the `STDOUT_CONSOLE` depend would need to remain on `!NATIVE_APPLICATION`
It would also make a lot of sense to have the libc cmake file not be included on that option instead of NATIVE_APP.

**Review by aescolar — CHANGES_REQUESTED** [review]
Thanks for this @vanwinkeljan . Some requests for improvements

File: drivers/counter/Kconfig.stm32_rtc (hunk 5 above)
Comment (vanwinkeljan) [reply drivers/counter/Kconfig.stm32_rtc]: >  I hope removing these selects will not cause too many issues to users

I hope the same  but that is the price we need to pay for using choice statement on the other hand the use of `select` is discouraged. 

> Including these select changes would discourage from merging during freeze even though this PR includes a bugfix for another issue.

Agree on this and think it is best to move this PR to 2.1

> Would it make sense to separate the select changes commit into a separate PR for after 2.0?

Short answer no, problem is that select does not work on choice symbols

```
warning: the choice symbol NEWLIB_LIBC (defined at lib/libc/Kconfig:21) is selected by the following symbols, but select/imply has no effect on choice symbols
 - CBOR_FLOATING_POINT (defined at modules/Kconfig.tinycbor:38)
 - CBOR_HALF_FLOAT_TYPE (defined at modules/Kconfig.tinycbor:44)
 - IEEE802154_KW41Z (defined at drivers/ieee802154/Kconfig.kw41z:9)
 - RTC_STM32 (defined at drivers/rtc/Kconfig.stm32_rtc:9)
 - COUNTER_RTC_STM32 (defined at drivers/counter/Kconfig.stm32_rtc:8)
 - UPDATEHUB (defined at lib/updatehub/Kconfig:7)
 - SIMPLELINK_HOST_DRIVER (defined at ext/hal/ti/simplelink/Kconfig:10)
```

**Comment by pfalcon** [comment]
> Added choice for c library impl. + corrected newlib dependencies 

There's something wrong with this PR, because Zephyr already offers a choice of c library impl.

**Comment by vanwinkeljan** [comment]
@pfalcon  
>There's something wrong with this PR, because Zephyr already offers a choice of c library impl.

Ok maybe it should read Kconfig choice

File: lib/libc/Kconfig [hunk 7]
```diff
@@ -7,18 +7,39 @@
...
 config NEWLIB_LIBC
-	bool "Build with newlib c library"
+	bool "Newlib c library"
+	depends on !NATIVE_APPLICATION
 	help
 	  Build with newlib library. The newlib library is expected to be
 	  part of the SDK in this case.
 
+config EXTERNAL_LIBC
+	bool "External c library"
```
Comment (aescolar) [inline lib/libc/Kconfig]: Very minor nit: Capital "C" in all these options descriptions.

File: drivers/counter/Kconfig.stm32_rtc (hunk 5 above)
Comment (aescolar) [reply drivers/counter/Kconfig.stm32_rtc]: > select does not work on choice symbols

@vanwinkeljan Yep. I guess only keeping them as independent options would be possible to not change the selects.

If you would prefer to not go that route for a quick 2.0 fix, then I guess it would be nice to tag this PR and the bug with milestone set to 2.1 (or future)

(Note that separating the select->depend into its own commit does not really help if it is not possible to revert it independently or bisect between them)

**Comment by vanwinkeljan** [comment]
@aescolar 

> (Note that separating the select->depend into its own commit does not really help if it is not possible to revert it independently or bisect between them)

If we follow this reasoning everything should get squashed into a single commit as none of them can stand on its own.

**Comment by aescolar** [comment]
Sidetracking:
> If we follow this reasoning everything should get squashed into a single commit as none of them can stand on its own.

In principle we should keep master bisectable (But i'm not going to be the one asking you to squash them back)

File: lib/libc/Kconfig [hunk 8]
```diff
@@ -76,10 +85,22 @@ config MINIMAL_LIBC_MALLOC_ARENA_SIZE
...
-	depends on !NEWLIB_LIBC
 	default y if 64BIT
 	help
 	  Build with long long printf enabled. This will increase the size of
 	  the image.
 
+endif #MINIMAL_LIBC
+
+
+config STDOUT_CONSOLE
```
Comment (pabigot) [inline lib/libc/Kconfig]: Remove double splacing

Comment (vanwinkeljan) [reply lib/libc/Kconfig]: done

**Comment by pabigot** [comment]
> In principle we should keep master bisectable (But i'm not going to be the one asking you to squash them back)

I did asked for them to be squashed, and still don't see any benefit to keeping them separate as they're pieces of the same task, but the blocker was for mixing in STM32 changes so I'm not insisting either.

**Review by pabigot — DISMISSED** [review]

**Review by aescolar — APPROVED** [review]
Ok, but for merging after 2.0 is released

**Comment by vanwinkeljan** [comment]
> I did asked for them to be squashed, and still don't see any benefit to keeping them separate as they're pieces of the same task, but the blocker was for mixing in STM32 changes so I'm not insisting either.

Ok I will squash all of them as none of them can stand on its own (including the STM32 changes), at least this will make the life of bisect easier.

**Comment by pabigot** [comment]
> Ok I will squash all of them as none of them can stand on its own (including the STM32 changes), at least this will make the life of bisect easier.

Please do **not** squash the STM32 changes.  Part of my confusion in reviewing this PR is that it is apparently (from other comments) fixing a bug related to STM32 and/or SimpleLink.  That is completely separate from changing the assumption that newlib is the only alternative to minimal libc, and was not at all clear in the commit messages.

**Comment by ioannisg** [comment]
@aescolar this fixes a bug - why not having it in 2.0?

**Comment by aescolar** [comment]
@ioannisg See:
https://github.com/zephyrproject-rtos/zephyr/pull/18315#discussion_r314591871 
It is changing quite a few kconfig options, in a way which may cause issues.
It is just a matter of the balance between the importance of the bug (low, it has always been there) vs the risk of breaking something during the stabilization phase.
But, you can take it in if you are ok with it

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by aescolar** [comment]
@vanwinkeljan Please rebase

**Review by MaureenHelm — APPROVED** [review]
@galak please review

File: tests/drivers/build_all/sensors_i_z.conf [hunk 9]
```diff
@@ -1,5 +1,6 @@
 CONFIG_TEST=y
 CONFIG_TEST_USERSPACE=y
+CONFIG_NEWLIB_LIBC=y
```
Comment (galak) [inline tests/drivers/build_all/sensors_i_z.conf]: what's going on here?

**Review by galak — COMMENTED** [review]
What's going on with adding CONFIG_NEWLIB_LIBC to the build_all tests?

File: tests/drivers/build_all/sensors_i_z.conf (hunk 9 above)
Comment (vanwinkeljan) [reply tests/drivers/build_all/sensors_i_z.conf]: STM hal is including math.h and minimal libc does not provided math.h

Comment (pabigot) [reply tests/drivers/build_all/sensors_i_z.conf]: I don't remember this change when I approved the PR.  I'd rather see minimal libc gain math.h that the HALs need, assuming it's not going to involve anything beyond the header.

Comment (vanwinkeljan) [reply tests/drivers/build_all/sensors_i_z.conf]: @pabigot 
>  I'd rather see minimal libc gain math.h that the HALs need, assuming it's not going to involve anything beyond the header.

Actually the HAL kconfig `HAS_STMEMSC` was hiding this  by using `imply NEWLIB_LIBC`.

Comment (pabigot) [reply tests/drivers/build_all/sensors_i_z.conf]: Hmpf.  Still, doesn't this change the behavior for all platforms?

Comment (vanwinkeljan) [reply tests/drivers/build_all/sensors_i_z.conf]: > Still, doesn't this change the behavior for all platforms?

Actually this PR does not change any behavior, new libc was already required but it was hidden from the `prj.conf` because it was selected (via `imply`) by `HAS_STMEMSC` by changing this `select` in a `depend` it becomes visible in the `pr.conf` file

Comment (pabigot) [reply tests/drivers/build_all/sensors_i_z.conf]: I guess this is more test infrastructure magic, because I only see five sensors that select `HAS_STMEMSC` so I would expect a behavior change in every other sensor that wouldn't normally be building with newlib.

I tried `sanitycheck -T samples/sensor` on a master branch with `Kconfig.st`'s newlib dependency removed and everything passed except `qemu_x86_long samples/sensor/thermometer/sample.sensor.thermometer` which failed because the test's exclude clause is incorrect (what is going on with CI these days, anyway?).

So I can't see how to confirm or deny that this is necessary, or whether there's another workaround.  I'll leave it to somebody else.

Comment (pabigot) [reply tests/drivers/build_all/sensors_i_z.conf]: Thanks.  Those also pass on master with the `imply NEWLIB_LIBC` removed.

Comment (vanwinkeljan) [reply tests/drivers/build_all/sensors_i_z.conf]: Hmm strange, just double checked myself against latest master (c2869c7e78cff4c643153a4f1b11971189fffa8b) with the `imply NEWLIB_LIBC` removed from  `config HAS_STMEMSC` and if I run `sanitycheck -N -i -T zephyr/samples/shields/x_nucleo_iks01a3` 2 out of 2 tests fails due to missing `math.h`

```
.../modules/hal/st/sensor/stmemsc/lis2dw12_STdC/driver/lis2dw12_reg.h:48:10: fatal error: math.h: No such file or directory
 #include <math.h>
          ^~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
```

Comment (pabigot) [reply tests/drivers/build_all/sensors_i_z.conf]: I have no explanation:
```
tirzah[281]$ sanitycheck -N -i -T samples/shields/x_nucleo_iks01a3
JOBS: 24
Cleaning output directory /mnt/nordic/zp/zephyr/sanity-out
Selecting default platforms per test case
Building testcase defconfigs...
Filtering test cases...
2 tests selected, 388 tests discarded due to filters
total complete:    2/   2  100%  failed:    0
2 of 2 tests passed with 0 warnings in 18 seconds
tirzah[282]$ git diff | cat
diff --git a/modules/Kconfig.st b/modules/Kconfig.st
index f2827fc548..7d5b8dd4d6 100644
--- a/modules/Kconfig.st
+++ b/modules/Kconfig.st
@@ -7,8 +7,6 @@ config HAS_STLIB
 
 config HAS_STMEMSC
 	bool
-	imply NEWLIB_LIBC
-	imply NEWLIB_LIBC_FLOAT_PRINTF
 
 if HAS_STMEMSC
 
tirzah[283]$ git one | head -1
* c2869c7e78 (HEAD -> upstream/master, r.upstream/master) [Mon Sep 16 20:19:58 2019 +0200] doc: reference: settings: list non-volatile storage backend
tirzah[284]$ 
```
but if I go into `sanity-out/frdm_k64f/samples/shields/x_nucleo_iks01a3/standard/test` and `ninja menuconfig` to disable `NEWLIB_LIBC` then I can see the problem.  So something unknown is still introducing the dependency.

However, at least I can reproduce it.  The following seems to be sufficient for the case I can reproduce:
```
#ifndef ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_MATH_H_
#define ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_MATH_H_

/* STMEMSC drivers require this header with this typedef. */
typedef float float_t;

#endif /*  ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_MATH_H_ */
```

**Review by galak — CHANGES_REQUESTED** [review]
Sorry for the delay on this.  I think we need to tweak this some so that we aren't having to update .conf files to set `CONFIG_NEWLIB_LIBC`.

I think we need a Kconfig option like 'COMPLETE_LIBC' or 'FULL_LIBC' or what major features of LIBC that a give bit of code might depend on.  Have the Kconfig's select that and than we can have a choice of which LIBC support 'FULL_LIBC' and default to NEWLIB.

**Comment by vanwinkeljan** [comment]
@galak I have no problem in adding  `COMPLETE_LIBC` or `FULL_LIBC` but is it a good idea to add the `select` back to the Kconfigs? By adding the `select` you are actually hiding the fact that you need a full libc implementation.

**Comment by vanwinkeljan** [comment]
I would rather see #19402 merged first add label blocked

**Comment by galak** [comment]
> @galak I have no problem in adding `COMPLETE_LIBC` or `FULL_LIBC` but is it a good idea to add the `select` back to the Kconfigs? By adding the `select` you are actually hiding the fact that you need a full libc implementation.

I think people are more wanting their driver for a given sensor to be enabled or such and not having to figure out that it happens to require newlib or a 'full-libc'.  I agree its not the best situation, but I think from a user point of view it makes more sense than the fact that I've got to enable newlib/full-libc to than even know that a bunch of drivers are available.

**Comment by vanwinkeljan** [comment]
> I think people are more wanting their driver for a given sensor to be enabled or such and not having to figure out that it happens to require newlib or a 'full-libc'. I agree its not the best situation, but I think from a user point of view it makes more sense than the fact that I've got to enable newlib/full-libc to than even know that a bunch of drivers are available. 

Maybe I'm an exception than but I rather see all the required configs in my top level project configuration than a driver just selecting one behind my back, especially if the config option has a rather large footprint.

**Comment by aescolar** [comment]
@galak, is what you are proposing to change @vanwinkeljan 's PR like this https://github.com/aescolar/zephyr/commit/c0d6f5e4364e5b534b7e9f2b446e4033e3026eb2 ?
(so the result is roughly this https://github.com/zephyrproject-rtos/zephyr/compare/master...aescolar:fix_newlib_native_posix )
If so, to me it would seem quite reasonable, as it would only be changing the default (where before it was going further and mandating `NEWLIB_LIBC`)

**Comment by galak** [comment]
> @galak, is what you are proposing to change @vanwinkeljan 's PR like this [aescolar@c0d6f5e](https://github.com/aescolar/zephyr/commit/c0d6f5e4364e5b534b7e9f2b446e4033e3026eb2) ?
> (so the result is roughly this [master...aescolar:fix_newlib_native_posix](https://github.com/zephyrproject-rtos/zephyr/compare/master...aescolar:fix_newlib_native_posix) )
> If so, to me it would seem quite reasonable, as it would only be changing the default (where before it was going further and mandating `NEWLIB_LIBC`)

Yes that looks like what I'm proposing.

**Comment by galak** [comment]
> > I think people are more wanting their driver for a given sensor to be enabled or such and not having to figure out that it happens to require newlib or a 'full-libc'. I agree its not the best situation, but I think from a user point of view it makes more sense than the fact that I've got to enable newlib/full-libc to than even know that a bunch of drivers are available.
> 
> Maybe I'm an exception than but I rather see all the required configs in my top level project configuration than a driver just selecting one behind my back, especially if the config option has a rather large footprint.

Like I said its not the best, but I'd rather we leave it behaving how it was.

**Comment by vanwinkeljan** [comment]
Updated according @galak proposal

@aescolar I based my self on your trial if you prefer to be added to the `Signed-off-by` let me know and I update the commit messages

**Comment by aescolar** [comment]
> @aescolar I based my self on your trial if you prefer to be added to the Signed-off-by let me know and I update the commit messages

It does not matter :)

**Review by aescolar — APPROVED** [review]

**Review by pabigot — APPROVED** [review]

File: drivers/sensor/grove/Kconfig [hunk 10]
```diff
@@ -8,7 +8,7 @@
 
 config GROVE_LIGHT_SENSOR
 	bool "Enable the Seeed Grove Light Sensor"
-	depends on ADC && NEWLIB_LIBC
+	depends on ADC && !MINIMAL_LIBC
```
Comment (galak) [inline drivers/sensor/grove/Kconfig]: why don't these become `FULL`?

Comment (vanwinkeljan) [reply drivers/sensor/grove/Kconfig]: Not clear what the exact question is, so 2 answers:

1. There is no `FULL_LIBC` only a `REQUIRES_FULL_LIBC` to indicate that a given config requires a full featured libc. The exact libc implementation can be set by the user (`newlib` or `external`)
2. The original code used `depends on` and not `select`, so just kept it backward compatible

**Comment by vanwinkeljan** [comment]
@galak Any further input?

File: drivers/sensor/grove/Kconfig (hunk 10 above)
Comment (galak) [reply drivers/sensor/grove/Kconfig]: I was asking why isn't this 'depends on ADC && REQUIRES_FULL_LIBC`?

Comment (vanwinkeljan) [reply drivers/sensor/grove/Kconfig]: `REQUIRES_FULL_LIBC` is a hidden attribute used to change the _default_ value of the `choice LIBC_IMPLEMENTATION` and can only be set via a `select` in another attribute (or via a defconfig).
As such `REQUIRES_FULL_LIBC` can not be used in `depends on` of attributes as it is not guaranteed it will be set even when _full_ libc is selected.

**Comment by pabigot** [comment]
Tagged dev-review: this blocks #18224 which we need with this to make progress on #19754

**Review by galak — APPROVED** [review]
